README.md
# `R/paramedic`: Predicting Absolute and Relative Abundance by Modeling Efficiency to Derive Intervals and Concentrations <img src="man/figures/paramedic-logo.png" align="right" width="165px"/>

[![Build Status](https://travis-ci.org/statdivlab/paramedic.svg?branch=master)](https://travis-ci.org/statdivlab/paramedic)
[![codecov](https://codecov.io/gh/statdivlab/paramedic/branch/master/graph/badge.svg?token=GnLFG7QNsh)](https://codecov.io/gh/statdivlab/paramedic)
[![License: BSD-3](https://img.shields.io/badge/License-BSD--3--Clause-yellow)](https://opensource.org/licenses/BSD-3-Clause)

**Software author:** [Brian Williamson](https://bdwilliamson.github.io/)

**Methodology authors:** [Brian Williamson](https://bdwilliamson.github.io/), [Jim Hughes](http://faculty.washington.edu/jphughes/) and [Amy Willis](http://statisticaldiversitylab.com/team/amy-willis)

------------------------------

## Introduction

`paramedic` is a R package for estimating microbial concentration. `paramedic` uses information from 16S count data (compositional data on all taxa) and absolute data on a subset of taxa (e.g., qPCR or flow cytometry) to estimate the absolute abundance of all taxa. The method accounts for differing taxon detection efficiencies between the two methods, and produces prediction and confidence intervals as well as point estimates of the absolute abundance of all taxa. Check out [the paper](https://www.biorxiv.org/content/10.1101/761486v1) for more details.

------------------------------

## R installation

You may install a stable release of `paramedic` from GitHub via  [`devtools`](https://www.rstudio.com/products/rpackages/devtools/) by running the following code (replace `v0.1.0` with the tag for the specific release you wish to install, and only install `devtools` if needed):

```r
install.packages("devtools")
devtools::install_github(repo = "statdivlab/paramedic@v0.1.0")
```

You may install a development release of `paramedic` from GitHub via [`devtools`](https://www.rstudio.com/products/rpackages/devtools/) by running the following code:

```r
install.packages("devtools")
devtools::install_github(repo = "statdivlab/paramedic")
```

------------------------------

## Example

This example shows how to use `paramedic` in a simple setting with data inspired by the University of Washington (UW) Sexually Transmitted Infections Cooperative Research Center (STICRC). For more examples and detailed explanation, please see the [vignette](https://statdivlab.github.io/paramedic/articles/introduction_to_paramedic.html).

```r
## load required functions and libraries
library("rstan")
library("paramedic")

## read in the data
data(example_16S_data)
data(example_qPCR_data)

# sample hyperparameter values
sigma_beta <- 1
sigma_Sigma <- 1
alpha_sigma <- 2
kappa_sigma <- 1

## -------------------------------------------------------------
## Estimate concentrations for the first 7 taxa
## -------------------------------------------------------------
## this is a small number of iterations, only for illustration
## also, shows how to use control parameters for rstan::stan
stan_mod <- run_paramedic(W = example_16S_data[, 1:8], V = example_qPCR_data,
                      sigma_beta = sigma_beta, sigma_Sigma = sigma_Sigma,
                      alpha_sigma = alpha_sigma, kappa_sigma = kappa_sigma,
                      n_iter = 30, n_burnin = 25, n_chains = 1, stan_seed = 4747,
                      control = list(adapt_delta = 0.85, max_treedepth = 15),
                      verbose = FALSE)
stan_mod_summ <- summary(stan_mod, probs = c(0.025, 0.975))$summary
stan_mod_samps <- extract(stan_mod)

## -------------------------------------------------------------
## Extract posterior estimates for the taxon missing abolute abundance data
## -------------------------------------------------------------
posterior_summaries <- extract_posterior_summaries(stan_mod_summ, stan_mod_samps,
                                                   taxa_of_interest = 9, mult_num = 1,
                                                   level = 0.95, interval_type = "wald")

## -------------------------------------------------------------
## Print out the posterior mean efficiencies and concentrations
## -------------------------------------------------------------
posterior_summaries$estimates
posterior_summaries$est_efficiency
```

------------------------------

## Citation

After using the `paramedic` package, please cite the following:

```
  @article{williamson2020,
    author={Williamson, BD and Hughes, JP and Willis, AD},
    title={A multi-view model for relative and absolute microbial abundances},
    journal={bioRxiv},
    year={2020},
    note={doi: https://doi.org/10.1101/761486}
  }
```

------------------------------

## Issues

We use Stan to fit the hierarchical model to the data. Read any warning messages returned by the algorithm carefully, as these can help diagnose convergence issues. Carefully choosing initialisation values can speed up convergence.

If you encounter any bugs or have any specific feature requests, please [file an issue](https://github.com/statdivlab/paramedic/issues). If your feature request relates to the hierarchical models implemented in `paramedic`, please read [the hierarchical models vignette](https://statdivlab.github.io/paramedic/articles/hierarchical_model_specification.html) prior to filing an issue or creating a pull request.

------------------------------


NEWS.md
# paramedic 0.1.3.5

## Major changes

None

## Minor changes

* Allow posterior draws of `e` to be used in predictions (with the flag `use_e = TRUE`, the new default in `posterior_predict`) or generate `e` based on posterior draws of `sigma_e` (obtained by setting `use_e = FALSE`)

# paramedic 0.1.3

## Major changes

* added a second option for obtaining predictions on new data using a previously fitted model: the options are now `predict_paramedic` and `posterior_predict`. The former uses posterior draws and a Stan algorithm to generate predictions; the latter uses posterior draws and internal R functions to generate predictions. Both are expected to perform similarly.

## Minor changes

* Updated return type of `run_paramedic` to class `paramedic`; added associated `print`, `summary` functions

# paramedic 0.1.2

## Major changes

* added ability to obtain predictions on new data using a previously fitted model using `predict_paramedic`

## Minor changes

None

# paramedic 0.1.1

## Major changes

* added ability to run analysis adjusted for batch effects by setting argument `k` to be nonzero in `run_paramedic` (the default is zero for data collected in a single experiment)
* added new stan code for batch analysis

## Minor changes

* reorganized stan code into directories, since much of the code was shared between models; the new organization uses `#include` statements, and the shared code is stored in `data/` (sets up data objects), `parameters/` (sets up model parameters), `tparameters/` (sets up transformed parameters), and `model/` (sets up hierarchical model for shared parameters)

# paramedic 0.1.0

## Major changes

* removed function `run_paramedic_centered`; the desired behavior can now be obtained by using `run_paramedic` with the argument `"centered = TRUE"`
* removed function `no_efficiency`; the desired behavior can now be obtained by using `run_paramedic` with arguments `"alpha_sigma = 0"` and `"kappa_sigma = 0"`
* removed several stan models by collapsing functionality into `variable_efficiency.stan` and `variable_efficiency_centered.stan`

## Minor changes

* updated all tests to reflect major changes
* added vignette discussing hierarchical model specification
* added a `pkgdown` site

# paramedic 0.0.3.900

## Major changes

None

## Minor changes

* new internal functions to reduce replicating code within key functions
* new `stan` model for specifying a negative binomial distribution on V

# paramedic 0.0.3

## Major changes

None

## Minor changes

* update internal `Stan` models to use log densities and unconstrained parameters, then back-transform after sampling

# paramedic 0.0.2

## Major changes

* allow an option to adjust for covariates

## Minor changes

None

# paramedic 0.0.1

## Major changes

None

## Minor changes

* allow `run_paramedic` with `q` = `q_obs`

# paramedic 0.0.0.9000

This is the initial package release!


R/run_paramedic.R
#' Run a Stan algorithm to estimate concentrations from combined absolute and relative abundance data
#'
#' Estimate concentrations (and efficiencies, if applicable) by combining absolute and relative abundance data.
#'
#' @param W The relative abundance data, e.g., from broad range 16S sequencing with "universal" primers. Expects data (e.g., matrix, data.frame, tibble) with sample identifiers in the first column. Sample identifiers must be the same between W and V, and the column must have the same name in W and V.
#' @param V The absolute abundance data, e.g., from taxon-specific absolute primers. Expects data (e.g., matrix, data.frame, tibble) with sample identifiers in the first column. Sample identifiers must be the same between W and V, and the column must have the same name in W and V.
#' @param X The covariate data. Expects data (e.g., matrix, data.frame, tibble) with sample identifiers in the first column. Sample identifiers must be the same between W, V, and X, and the column must have the same name in W, V, and X. If X only consists of the subject identifiers, then no covariates are used.
#' @param k the number of batches that the relative abundance data W were analyzed in. If k = 0 (the default), then batch effects are not considered.
#' @param n_iter The total number of iterations per chain to be run by the Stan algorithm. Defaults to 10500.
#' @param n_burnin The total number of warmups per chain to be run by the Stan algorithm. Defaults to 10000.
#' @param n_chains The total number of chains to be run by the Stan algorithm. Defaults to 4.
#' @param stan_seed The random number seed to initialize.
#' @param centered If \code{TRUE}, uses a centered parameterization on the true concentrations \eqn{\mu}; if \code{FALSE} (the default), uses a non-centered parameterization.
#' @param inits_lst An optional list of initial values of the parameters. Must be a named list; see \code{\link[rstan]{stan}}.
#' @param sigma_beta Hyperparameter specifying the prior variance on \eqn{\beta_0}. Defaults to \eqn{\sqrt{50}}.
#' @param sigma_Sigma Hyperparameter specifying the prior variance on \eqn{\Sigma}. Defaults to \eqn{\sqrt{50}}.
#' @param alpha_sigma Hyperparameter specifying the shape parameter of the prior distribution on \eqn{\sigma_e}. Defaults to 2.
#' @param kappa_sigma Hyperparameter specifying the scale parameter of the prior distribution on \eqn{\sigma_e}. Defaults to 1.
#' @param alpha_phi Hyperparameter specifying the shape parameter of the prior distribution on \eqn{\phi}. Defaults to 0; a negative binomial model can be specified if both \code{alpha_phi} and \code{beta_phi} are nonzero.
#' @param beta_phi Hyperparameter specifying the rate parameter of the prior distribution on \eqn{\phi}. Defaults to 0; a negative binomial model can be specified if both \code{alpha_phi} and \code{beta_phi} are nonzero.
#' @param sigma_xi Hyperparameters specifying the variance of efficiencies over batches. Only used if \code{k} is greater than zero. Defaults to 1.
#' @param ... other arguments to pass to \code{\link[rstan]{sampling}} (e.g., control).
#'
#' @return An object of class \code{stanfit}.
#'
#' @details We fit a hierarchical model in Stan to the data, with goal to estimate true concentration for all taxa. There are two available hierarchical models. The two only differ in the way that the true concentration is parameterized. While these hierarchical models are mathematically identical, they are fit differently by Stan. The first hierarchical model is a "centered" model, where \deqn{\beta_0 ~ N(0, \sqrt{50})} \deqn{\log \Sigma ~ N(0, \sqrt{50})} \deqn{\log \mu ~ N(\beta_0 + X\beta_1, \Sigma).} We call this model \code{variable_efficiency_centered.stan}. The second hierarchical model is a "noncentered" model, where \deqn{\beta_0 ~ N(0, \sqrt{50})} \deqn{\log \Sigma ~ N(0, \sqrt{50})} \deqn{\log \gamma ~ N(0, I)} \deqn{\log \mu = \sqrt{\Sigma} \log \gamma + \beta_0 + X\beta_1.} We call this model \code{variable_efficiency.stan}.

#' In most cases, we suggest using the noncentered model. However, if the model does not converge in a reasonable amount of time using the noncentered model, consider trying the centered model.
#'
#' @examples
#' # load the package, read in example data
#' library("paramedic")
#' data(example_16S_data)
#' data(example_qPCR_data)
#'
#' # run paramedic (with an extremely small number of iterations, for illustration only)
#' # on only the first 10 taxa
#' mod <- run_paramedic(W = example_16S_data[, 1:10], V = example_qPCR_data,
#' n_iter = 30, n_burnin = 25, n_chains = 1, stan_seed = 4747)
#'
#' @seealso \code{\link[rstan]{stan}} and \code{\link[rstan]{sampling}} for specific usage of the \code{stan} and \code{sampling} functions.
#'
#' @export
run_paramedic <- function(W, V, X = V[, 1, drop = FALSE], k = 0,
                      n_iter = 10500, n_burnin = 10000, n_chains = 4, stan_seed = 4747,
                      centered = FALSE, inits_lst = NULL,
                      sigma_beta = sqrt(50), sigma_Sigma = sqrt(50), alpha_sigma = 2, kappa_sigma = 1, 
                      alpha_phi = 0, beta_phi = 0, sigma_xi = 1,
                      ...) {
    # --------------
    # error messages
    # --------------
    check_entered_data(W, V, X, k, inits_lst, sigma_beta, sigma_Sigma, alpha_sigma, kappa_sigma)
    # ---------------------------
    # pre-processing and warnings
    # ---------------------------
    pre_processed_lst <- make_paramedic_tibbles(W, V, X, k, inits_lst, sigma_beta, sigma_Sigma, alpha_sigma, kappa_sigma)
    W_mat <- pre_processed_lst$w_mat
    V_mat <- pre_processed_lst$v_mat
    X_mat <- pre_processed_lst$x_mat
    # ----------------------------------------
    # set up the data and initial values lists
    # ----------------------------------------
    data_inits_lst <- make_paramedic_stan_data(k, W_mat, V_mat, X_mat, inits_lst, 
                                               sigma_beta, sigma_Sigma, 
                                               alpha_sigma, kappa_sigma, 
                                               alpha_phi, beta_phi,
                                               n_chains, centered)
    data_lst <- data_inits_lst$data_lst
    inits_lst <- data_inits_lst$inits_lst
    # ----------------------
    # run the Stan algorithm
    # ----------------------
    pars <- c("mu",
              if (alpha_sigma > 0 & kappa_sigma > 0) "e",
              "beta_0",
              if (ncol(X_mat) > 0) "beta_1",
              "Sigma",
              if (alpha_phi > 0 & beta_phi > 0) "phi",
              if (alpha_sigma > 0 & kappa_sigma > 0) "sigma_e")
    stan_model <- switch((k > 0) + 1, 
                         switch(centered + 1,
                         stanmodels$variable_efficiency,
                         stanmodels$variable_efficiency_centered),
                         stanmodels$variable_efficiency_batches)
    mod <- rstan::sampling(stan_model, data = data_lst, pars = pars,
                           chains = n_chains, iter = n_iter, warmup = n_burnin, seed = stan_seed,
                           init = inits_lst, ...)
    output <- list(
        stan_fit = mod,
        stan_data = data_lst,
        stan_inits = inits_lst,
        stan_model = stan_model,
        summary = rstan::summary(mod, probs = c(0.025, 0.975))$summary
    )
    structure(output, class = "paramedic")
}
